### PR TITLE
Fixes for PutObjDirect and default values for amazon object client env vars

### PR DESCRIPTION
**Pull request by Tryneus**
Added a new test for putting and reading an object with `DirectObjWriter` and `DirectObjReader`, uncovered a bug where we would double-write data if there was only one chunk of data.  Also found out that we weren't actually getting default values for the `AmazonAdvancedConfiguration` fields because they included a space in their tag.  Rather than remove the space, I changed the tag parser to account for this because it's not obviously wrong and reads better with the space.  Could maybe use more generic tag parsing, but that's a bit out of scope.  Left in a few `errors.EnsureStack` calls for errors that obviously are missing stack traces.

### CODE REVIEW

File: src/server/pkg/obj/obj.go [hunk 1]
```diff
@@ -132,7 +132,7 @@ type AmazonAdvancedConfiguration struct {
 	// uploader, and not the owner of the bucket. Using the default ensures that
 	// the owner of the bucket can access the objects as well.
 	UploadACL      string `env:"UPLOAD_ACL, default=bucket-owner-full-control"`
-	Reverse        bool   `env:"REVERSE, default=true"`
+	Reverse        bool   `env:"REVERSE, default=false"`
```
Comment (Tryneus) [inline src/server/pkg/obj/obj.go:135]: I flipped this from 'true' to 'false' because this default has been 'false' in practice due to the silent discarding of the default value.

Comment (jdoliner) [inline src/server/pkg/obj/obj.go:135]: I think we want to keep this as default=true as lame as that is. Because otherwise existing user's clusters will break when they upgrade and use the same `pachctl deploy` command that they used in the previous version.

File: src/client/pfs.go [hunk 2]
```diff
@@ -1872,7 +1872,7 @@ func (w *putObjWriteCloser) Write(p []byte) (int, error) {
 }
 
 func (w *putObjWriteCloser) Close() error {
-	if w.request.Obj == "" {
+	if w.request.Obj != "" {
```
Comment (jdoliner) [inline src/client/pfs.go:1875]: I'm a little unsure if this change is correct. Based on the comment below it looks like this line is supposed to run when the block is empty, which I think is the same as when `Obj == ""`. Not sure though, I'd defer to @brycemcanally on this.

File: src/server/pkg/obj/obj.go (hunk 1 above)
Comment (Tryneus) [reply src/server/pkg/obj/obj.go:135]: I don't see why we would keep this as `default=true` - the old default was never used until the fix in `cmdutil/env.go` in this PR.  The old default in practice was `false`.

File: src/client/pfs.go (hunk 2 above)
Comment (Tryneus) [reply src/client/pfs.go:1875]: To be clear, this `putObjWriteCloser` is only used for the `PutObjDirect` RPC, which until 1.11.5 was not used anywhere in the codebase (it was used a bit in the worker before the refactor, I believe, but that went away until I changed the job artifact object usage in 1.11.5).  `Obj == ""` appears to only be checking if a previous chunk of data has been sent.  In practice, this logic being wrong meant that we would send an extra unused message at the end of the stream, and writing an empty object would error.  I've added another test that reproduces this with an empty object on the old code and works fine with the new code.

Comment (jdoliner) [reply src/client/pfs.go:1875]: Ahh, that makes sense.

File: src/server/pkg/obj/obj.go (hunk 1 above)
Comment (jdoliner) [reply src/server/pkg/obj/obj.go:135]: Interesting I didn't realize this was false in practice, that's actually kind of convenient.

**Review by jdoliner — APPROVED** [review]
LGTM
